### PR TITLE
Retrieve and store settings from/to local storage.

### PR DESCRIPTION
**Pull request by bmd3k**
* Add effects class that coordinates the retrieval and saving of settings properties to local storage.
* Add a data source layer that abstracts away the local storage interactions. 
* Modify the reducer layer to update state based on the results.
* Modify the initial state so that it now says settings are NOT_LOADED. The Settings are now only considered LOADED when the attempt to fetch settings from local storage completes (either successfully or unsuccessfully).

Besides writing tests for each layer I also tested the flow on a live tensorboard by modifying settings and refreshing and seeing that my settings have been restored. I used the Chrome developer console to inspect local storage and manipulate properties.

### CODE REVIEW

File: tensorboard/webapp/settings/_redux/settings_effects_test.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    expect(saveReloadPeriodInMs).toHaveBeenCalledWith(1111);
+  });
+
+  it('saves reloadPeriodInMs', () => {
+    const savePageSize = spyOn(dataSource, 'savePageSize');
+
+    store.overrideSelector(getPageSize, 1111);
+    store.refreshState();
+    actions.next(changePageSize({size: 2222}));
+    expect(savePageSize).toHaveBeenCalledWith(1111);
```
Comment (japie1235813) [inline tensorboard/webapp/settings/_redux/settings_effects_test.ts:130]: Not sure if I understand this. Why `savePageSize` method is called with `1111` not `2222`?

File: tensorboard/webapp/settings/_data_source/settings_data_source.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      SETTINGS_RELOAD_ENABLED_KEY
+    );
+    if (reloadEnabled) {
+      savedSettings.reloadEnabled = reloadEnabled === 'true';
+    }
+    const reloadPeriodInMs = window.localStorage.getItem(
+      SETTINGS_RELOAD_PERIOD_IN_MS_KEY
+    );
+    if (reloadPeriodInMs) {
+      savedSettings.reloadPeriodInMs = parseInt(reloadPeriodInMs);
```
Comment (japie1235813) [inline tensorboard/webapp/settings/_data_source/settings_data_source.ts:38]: Internal lint tools would complain about this. Use Number(reloadPeriodInMs) or parseInt(reloadPeriodInMs, 10) (specify the redix)

Comment (bmd3k) [reply tensorboard/webapp/settings/_data_source/settings_data_source.ts]: Done.

File: tensorboard/webapp/settings/_redux/settings_effects.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import {Injectable} from '@angular/core';
+import {Actions, createEffect, ofType, OnInitEffects} from '@ngrx/effects';
+import {Action, createAction, Store} from '@ngrx/store';
+import {EMPTY} from 'rxjs';
+import {catchError, switchMap, tap, withLatestFrom} from 'rxjs/operators';
+
+import {State} from '../../app_state';
+import {SettingsDataSource} from '../_data_source/settings_data_source';
+import * as actions from './settings_actions';
+import * as selectors from './settings_selectors';
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/settings/_redux/settings_effects.ts]: Done.

File: tensorboard/webapp/settings/_redux/settings_effects.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            actions.fetchSavedSettingsSucceeded({savedSettings})
+          );
+        }),
+        catchError(() => {
+          this.store.dispatch(actions.fetchSavedSettingsFailed());
+          return EMPTY;
+        })
+      );
+    },
+    {dispatch: false}
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/webapp/settings/_redux/settings_effects.ts:74]: `dispatch: false` indicates the corresponding observable will not return an action to be dispatched. It does not mean we will not dispatch our own actions while processing/generating the observable.

https://ngrx.io/api/effects/createEffect

This is a common pattern in our code base:

https://cs.opensource.google/search?q=%22dispatch:%20false%22&ss=tensorflow%2Ftensorboard

File: tensorboard/webapp/settings/_redux/settings_effects_test.ts (hunk 1 above)
Comment (bmd3k) [reply tensorboard/webapp/settings/_redux/settings_effects_test.ts]: Good question. Added comments. Let me know if it needs more clarity.

**Pull request closed without merging**